### PR TITLE
Write custom Function and gradcheck

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67918
* #67917
* __->__ #67916

### CODE REVIEW

File: test/test_autograd_onboarding_lab.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            ''' Inputs
+                q: m * n
+                k: m * n
+                v: m * p
+                Outputs
+                a: m * m
+                o: m * p
+            '''
+            # check that inputs are 2D and check sizes of input
+            assert q.dim() == 2 and k.dim() == 2 and v.dim() == 2, "all inputs must have dim=2"
```
Comment (albanD) [inline test/test_autograd_onboarding_lab.py:19]: Good to do these checks!

File: test/test_autograd_onboarding_lab.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            assert q.size()[1] == k.size()[1], "1st dim of q and 1st dim of k must have the same size"
+            assert k.size()[0] == v.size()[0], "0th dim of k and 0th dim of v must have the same size"
+
+            x = torch.matmul(q, k.transpose(0, 1))
+            a = torch.tanh(x)
+            o = torch.matmul(a, v)
+            ctx.save_for_backward(q, k, v, a)
```
Comment (albanD) [inline test/test_autograd_onboarding_lab.py:26]: Note that you don't have to save intermediary results in this way, only input/output need to be here.
But for simplicity it's ok to treat all of them the same way here.

File: test/test_autograd_onboarding_lab.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    def test_onboarding_aten(self):
+        M = 20
+        N = 30
+        P = 15
+        attn = self.Onboarding_Attn.apply
+        input = (torch.randn(M, N, dtype=torch.double, requires_grad=True),
+                 torch.randn(M, N, dtype=torch.double, requires_grad=True),
+                 torch.randn(M, P, dtype=torch.double, requires_grad=True))
+        test_gradcheck = gradcheck(attn, input, eps=1e-6, atol=1e-4)
```
Comment (albanD) [inline test/test_autograd_onboarding_lab.py:64]: Did you have to change the atol here?

File: test/test_autograd_onboarding_lab.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        test_gradgradcheck = gradgradcheck(attn, input, eps=1e-6, atol=1e-4)
+        assert test_gradcheck, "Gradcheck failed"
```
Comment (albanD) [inline test/test_autograd_onboarding_lab.py:66]: The functions above will already raise errors by default when they fail.
Also within the test suite, you should use `self.assertFoo()`.

**Review by albanD — APPROVED** [review]
Looks good. Just some comments

File: test/test_autograd_onboarding_lab.py (hunk 2 above)
Comment (mikaylagawarecki) [reply test/test_autograd_onboarding_lab.py:26]: Just to make sure I'm understanding, I should save q, k, v, a, o here because they are the inputs + outputs and x should not be saved here because gradients will not flow in backward, is that right?

Comment (albanD) [reply test/test_autograd_onboarding_lab.py:26]: double backward will indeed be problematic for x because gradient won't flow properly.
